Autoconfirm XMR trade if tx check was validated by proof service

jmacxx opened the pull request:
This is a follow-up PR to #4378 by @chimp1984

Implements bisq-network/proposals#86


### Features:

* XMR seller is prompted to enter txId and viewkey.

On the XMR buyer side,
* looks up the XMR transaction to verify that it has confirmed (using [onion-monero-blockchain-explorer](https://github.com/moneroexamples/onion-monero-blockchain-explorer))
* user can run their own validating service, or use the ones provided
* 100% agreement of all chosen services is necessary for confirmation
* feature can be configured and turned on/off from settings screen
* feature can be globally turned off via admin filter
 

### Implementation notes

Auto confirmation entails API requests to multiple autoconf service providers.  If the responses all show that the required number of confirmations have been reached the trade will be auto confirmed.

The user still has the option to confirm manually just as if auto confirm were not enabled.

Any service failure will abort the attempt at auto confirming.
- if a service is down, we fail with CONNECTION_FAIL
- if a service returns garbage response, we fail with API_INVALID
- if a service returns a different address, we fail with ADDRESS_INVALID
- if a service returns a different txId, we fail with TX_HASH_INVALID
- if a service returns a different viewkey, we fail with TX_KEY_INVALID
- if a service returns a match, but the amount is different we fail with AMOUNT_NOT_MATCHING
- if a service returns transaction not found we wait then re-try
- if a service returns less than the required confirm count, we wait then re-try
- if a service returns >= the required confirm count, TradeManager decrements its count of outstanding service calls.  Once the count reaches zero it transitions the trade to confirmed.

In any of these cases the status is shown on screen.  It also shows the progress of the confirmation count.

The general approach to implementing this feature was to keep it as simple as possible.  

Auto confirmation is a one-shot process triggered by receipt of `counterCurrencyExtraData` in the `CounterCurrencyTransferStartedMessage` from the counterparty.  

If Bisq is closed while auto confirmation was pending, when it re-starts the trade will show auto conf status disabled.  
~~Auto-confirmed status of a trade is not persisted.~~ Auto-confirmed status is persisted since for trade history it might be important to have some "proof" if it was used or not, as well in dispute cases. [requested by @chimp1984].


### Testing

In regtest use a clearnet service address, e.g. `78.47.61.90:8081` (from @Emzy)

In mainnet over tor this would be `monero3bec7m26vx6si6qo7q7imlaoz45ot5m2b5z2ppgoooo6jx2rqd.onion` which is included as a default setting.

To try out this feature without making a live XMR trade, the following example TX proofs may be used.

    Sent 0.12121212 XMR (for sell 0.0012 BTC @ price 0.0099)
    to:  4ATyxmFGU7h3EWu5kYR6gy6iCNFCftbsjATfbuBBjsRHJM4KTwEyeiyVNNUmsfpK1kdRxs8QoPLsZanGqe1Mby43LeyWNMF
    tx:  0d0d5e74e56c9060095e62b92bce2c7fe8da43a28d3e11a7c51ffe0749fe4dd4
    key: 443263d31c8b05f5dd5734128fcb3cac83d0fd3c90b03ec19d5131216a5e3505

    Sent 0.1 XMR (for sell 0.001 BTC @ price 0.01)
    to:  4ATyxmFGU7h3EWu5kYR6gy6iCNFCftbsjATfbuBBjsRHJM4KTwEyeiyVNNUmsfpK1kdRxs8QoPLsZanGqe1Mby43LeyWNMF
    tx:  e8dcd8160aee016d8a0d9c480355d65773dc577313a0af8237c35f9d997b01c0
    key: 300fa18ff99b32ff097d75c64d62732bdb486af8c225f558ee48c5f777f9b509


Automated testing routines were created to verify the API JSON response handling.

Checking iteroperability between new and old clients.
- If BTC Seller = new BTC buyer = old, the buyer is not prompted for txId and key and the new proto field is not sent to the BTC seller.  Therefore auto-confirmation will not be started and the trade will operate in the old way, manual confirmation.
- If BTC Seller = old BTC buyer = new, the buyer provides txId and key and sends the new proto field to the BTC seller.  The BTC seller does not handle the new field and the trade will operate in the old way, manual confirmation.

**Test cases**

- New user (preferences)
- Auto-confirm success
- Premature manual confirm by user
- XMR Tx Id / key not provided
- XMR Tx not found
- Re-used XMR TxId
- Feature disabled in settings
- Feature disabled by filter
- Transaction limit exceeded
- Invalid amount
- XMR Trade date too old
- XMR Trade date newer than Bisq trade date (ok)
- Bisq shut down during Auto-confirmation procedure
- XMR API service not reachable




### Discussion

* Perhaps some XMR sellers may not want (or be able) to provide the txId and viewkey?  In interest of flexibility, the UI allows those values to be optional and simply disables autoconf for the trade if the keys are empty or do not pass validation.

chimp1984 committed:
Add basic support for validation for XMR transfer with tx key

Main part missing is the XMR proof service request processing. I did not
get the service compiled yet, so could not test response data and error
conditions.

Further it is missing a "news badge" and popup to guide the user to the
new feature.

Only basic dev tested so far.

Anyone welcome to pick the project up from here as I might not have
time soon to continue.

chimp1984 committed:
Add trade date

Commit:
Implement XMR tx proof autoconfirm feature

* XMR seller is prompted to enter txId and viewkey.
* looks up the XMR transaction to verify that it has confirmed
* user can run their own validating service, or use the ones provided
* 100% agreement of all chosen services is necessary for confirmation
* feature can be configured and turned on/off from settings screen
* feature can be globally turned off via admin filter
* two code review passes from chimp1984
* one text review from m52go

chimp1984 commented:
@jmacxx 
Great to see the final PR! I will try to review on the weekend as soon I find time.

> Perhaps some XMR sellers may not want (or be able) to provide the txId and viewkey?

It is a requirement for XMR that they use a wallet capable of providing the tx key for the BTC buyer to be able to proof the XMR transfer. In a dispute the mediator would not be able to verify if the XMR transfer was done otherwise. So we just make it more safe in the way that if user ignore the info in the popup and use a not capable wallet they would risk to lose the dispute case. Now they cannot complete the trade, or at least they get aware of the problem very early.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrProofInfo.java:
```diff
@@ -0,0 +1,152 @@
+            // TODO check if date of tx is after tradeDate (allow some tolerance to avoid clock sync issues).
+            //  Otherwise a scammer could send an old txKey from a prev trade with same amount before seller has
+            //  updated to new feature, thus the check for duplication would not detect the scam.
+            //  >>>>> the API does not provide date of tx <<<<<<
```
I think its just some edge case at update where this could be a risk. Does the API not provide the date? If not would you mind to ask the Monero dev if he can add it? I guess it would be a small change and would remove the described risk.

chimp1984 commented:
@boring-cyborg Wrong labelling!

chimp1984 commented on assets/src/main/java/bisq/asset/CryptoNoteAddressValidator.java:
```diff
@@ -57,6 +60,19 @@ public AddressValidationResult validate(String address) {
     }
+
+    // obtain the raw hex version of a monero address
+    public static String addressToHex(String address) {
```
I would suggest to name it more specifically. As we remove type and checksum it is not really converting but we use it for verifying if it is a valid XMR address. I think the hex check might be good enough for now but maybe there is a more exact validation available. If we rename the method to 'isAddressValid' and return a boolean I think it gets more clean.

chimp1984 commented on core/src/main/java/bisq/core/filter/Filter.java:
```diff
@@ -109,6 +109,9 @@
     private final List<String> btcFeeReceiverAddresses;
 
+    // added after v1.3.7
+    private final boolean disableAutoconf;
```
I would suggest to stick with camelCase -> `disableAutoConf`

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrProofInfo.java:
```diff
@@ -0,0 +1,152 @@
+
+    public XmrProofResult checkApiResponse(String jsonTxt) {
+        try {
+            final JsonObject json = new Gson().fromJson(jsonTxt, JsonObject.class);
```
We don't use `final` for local variables.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrProofInfo.java:
```diff
@@ -0,0 +1,152 @@
+            if (jsonAddress == null) {
+                return new XmrProofResult(XmrProofResult.State.API_INVALID, "missing address field");
+            } else {
+                String expectedAddressHex = CryptoNoteAddressValidator.addressToHex(this.recipientAddress);
```
Ah it seems you need the trimmed address for the comparision, if so ignore my comment earlier. Still might be good to have a validate method and a convert method to make it more explicit.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrProofInfo.java:
```diff
@@ -110,6 +117,20 @@ public XmrProofResult checkApiResponse(String jsonTxt) {
+                return new XmrProofResult(XmrProofResult.State.API_INVALID, "missing tx_timestamp field");
+            } else {
+                long tradeDateSeconds = tradeDate.getTime() / 1000;
+                long difference = abs(jsonTimestamp.getAsLong() - tradeDateSeconds);
```
I think the `abs` call here is leading to undesired behaviour. We only want to protect against old txs, so then the trade data  > tx_timestamp. If we allow both directions we limit the confirmation time to 1 day, which might be mostly ok for XMR but not sure... E.g. if tradeDateSeconds is today 18:00 and tx_timestamp is tomorrow 19:00 because miner fee was low and tx got included very late it would fail, but should be still valid.
I think correct impl would be:
`long difference = tradeDateSeconds - jsonTimestamp.getAsLong();
    if (difference > 60 * 60 * 24) {....
`
Not sure if jsonTimestamp is from confirmed block or from mempool, but should not make a difference.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrProofResult.java:
```diff
@@ -17,15 +17,81 @@
+        AMOUNT_NOT_MATCHING;
+    }
+
+    @Getter
```
Annotating the class with @Value would make it more explicit that its a immutable class. @Getter annotations can be removed then as @Value included those.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrProofResult.java:
```diff
@@ -17,15 +17,81 @@
+        switch (state) {
+            case TX_NOT_CONFIRMED:
+                return Res.get("portfolio.pending.autoConfirmPending")
+                        + " " + Integer.toString(confirmCount)
```
Integer.toString is not needed.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrTransferProofRequester.java:
```diff
@@ -17,158 +17,109 @@
+public class XmrTransferProofRequester {
 
     private final ListeningExecutorService executorService = Utilities.getListeningExecutorService(
             "XmrTransferProofService", 3, 5, 10 * 60);
```
Better use class name 'XmrTransferProofRequester'

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrTransferProofRequester.java:
```diff
@@ -17,158 +17,109 @@
+            return;
+        }
         ListenableFuture<XmrProofResult> future = executorService.submit(() -> {
             Thread.currentThread().setName("XmrTransferProofRequest-" + this.toString());
```
this.toString() would not result it any readable meaningful data (shows java internal id). Also if someone adds a toString mehtod we would get all that into the thread name rendering it unreadable...
Better use the tx hash + service url, that should be unique enough

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrTransferProofRequester.java:
```diff
@@ -17,158 +17,109 @@
-    //todo dev settings
-    private long REPEAT_REQUEST_SEC = TimeUnit.SECONDS.toMillis(5);
+    // these settings are not likely to change and therefore not put into Config
+    private long REPEAT_REQUEST_PERIOD = TimeUnit.SECONDS.toMillis(30);
```
Is 30 seconds not a bit too aggressive for the nodes? Not sure whats the best value here. What are average block confirmation times on XMR? I guess some value in between 1-3 min. might be better.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrTransferProofRequester.java:
```diff
@@ -17,158 +17,109 @@
+                    log.warn("we have tried this service API for too long, giving up: {}", httpClient.toString());
+                    return;
+                }
+                if (result.isPendingState())
```
Better use alwqays curly brackets also for one-liners. We had some bugs related to that in the past and changed then code style (but still in the code from old days).

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrTransferProofRequester.java:
```diff
@@ -17,158 +17,109 @@
+        this.httpClient = new XmrTxProofHttpClient(socks5ProxyProvider);
+        this.httpClient.setBaseUrl("http://" + xmrProofInfo.getServiceAddress());
+        if (xmrProofInfo.getServiceAddress().matches("^192.*")) {
+            log.info("ignoring Socks5 proxy for local net address: {}", xmrProofInfo.getServiceAddress());
```
I think there is another use case where we dont want to use the onion address. For regtest we usually dont use tor, so for dev testing its more convenient to use the clearnet IP of a service (at least one should provide clearnet access for that reason).

chimp1984 replied:
See ProvidersRepository how it uses @Named(Config.USE_LOCALHOST_FOR_P2P) boolean useLocalhostForP2P to determine if a clearnet address need to be used.

chimp1984 replied:
The regex with `^192.*` would not match if a user sets `localhost`.

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrTransferProofService.java:
```diff
@@ -17,65 +17,69 @@
+        XmrTransferProofRequester requester = map.getOrDefault(key, null);
+        if (requester != null) {
+            log.info("Terminating API request for {}", key);
+            requester.setTerminated(true);
```
I would find it better to call a method like stop() where the property is set instead of exposing the property itself. Small issue but makes API of the class more clear IMO.

chimp1984 commented on core/src/main/java/bisq/core/trade/Trade.java:
```diff
@@ -465,6 +475,7 @@ protected Trade(Offer offer,
         processModel = new ProcessModel();
         lastRefreshRequestDate = takeOfferDate;
         refreshInterval = Math.min(offer.getPaymentMethod().getMaxTradePeriod() / 5, MAX_REFRESH_INTERVAL);
+        xmrProofResult = new XmrProofResult(0, 0, XmrProofResult.State.FEATURE_DISABLED);
```
Why is xmrProofResult set here? It should be marked as a @Nullable fields and left null until used. Also we should use the setXmrProofResult method only to avoid that the xmrProofResultProperty gets out or sync.

chimp1984 commented on core/src/main/java/bisq/core/trade/Trade.java:
```diff
@@ -428,12 +429,21 @@ public Phase getPhase() {
+    }
+    @Getter
+    // This observable property can be used for UI to show a notification to user of the XMR proof status
+    transient final private ObjectProperty<XmrProofResult> xmrProofResultProperty = new SimpleObjectProperty<>(xmrProofResult);
```
I think we should not pass xmrProofResult here as it is null at construction time.

chimp1984 commented on core/src/main/java/bisq/core/trade/Trade.java:
```diff
@@ -1170,6 +1181,7 @@ public String toString() {
                 ",\n     takerPaymentAccountId='" + takerPaymentAccountId + '\'' +
                 ",\n     errorMessage='" + errorMessage + '\'' +
                 ",\n     counterCurrencyTxId='" + counterCurrencyTxId + '\'' +
+                ",\n     counterCurrencyExtraData='" + counterCurrencyExtraData + '\'' +
```
We could add xmrProofResult as well

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeManager.java:
```diff
@@ -932,65 +935,118 @@ public void processCounterCurrencyExtraData(Trade trade) {
+        return failure;
+    }
+
+    private boolean isDisableAutoconf() {
```
Maybe rename to `isAutoConfDisabledByFilter` to make it more distinct to preferences value.

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeManager.java:
```diff
@@ -932,65 +935,118 @@ public void processCounterCurrencyExtraData(Trade trade) {
+        if (!walletsSetup.hasSufficientPeersForBroadcast()) {
+            return failure;
+        }
+        if (!walletsSetup.isDownloadComplete()) {
```
I am not sure if that handling is ok (same for hasSufficientPeersForBroadcast). I guess the message which triggers the processCounterCurrencyExtraData can be received while user is syncing wallet and we dont want to stop the service in such a case but delay the requests. As it cannot be handled here I would suggest to move the 
```
 if (!walletsSetup.hasSufficientPeersForBroadcast()) {
            return failure;
        }
        if (!walletsSetup.isDownloadComplete()) {
            return failure;
        }
```
code to the 
```
result -> {
                            if (!handleProofResult(result, trade))
                                xmrTransferProofService.terminateRequest(xmrProofInfo);
                        }
```
part of the calling method.

But not sure if that is really enough as it need to be handled in other places as well where the result gets processed. Need to get a bit more overview to be able to make more meaningful suggestions....

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeManager.java:
```diff
@@ -932,65 +935,118 @@ public void processCounterCurrencyExtraData(Trade trade) {
-                                    }
+        // here we count the Trade's API results from all
+        // different serviceAddress and figure out when all have finished
+        Integer resultsCountdown = txProofResultsPending.getOrDefault(trade.getId(), 0);
```
We dont use boxed types if not needed.

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeManager.java:
```diff
@@ -932,65 +935,118 @@ public void processCounterCurrencyExtraData(Trade trade) {
+        }
+
+        if (result.isPendingState()) {
+            log.info("Received a {} message for tradeId {}, retry will happen automatically",
```
This log is too generic. it should contain more specific info about the actual use case/feature.

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeManager.java:
```diff
@@ -932,65 +935,118 @@ public void processCounterCurrencyExtraData(Trade trade) {
+            trade.setXmrProofResult(result);            // this updates the GUI with the status..
+            log.info("Auto confirm was successful, transitioning trade {} to next step...", trade.getShortId());
+            if (!trade.isPayoutPublished()) {
+                trade.setState(Trade.State.SELLER_CONFIRMED_IN_UI_FIAT_PAYMENT_RECEIPT);
```
Maybe we should add a comment here and to the `Trade.State.SELLER_CONFIRMED_IN_UI_FIAT_PAYMENT_RECEIPT` state that this can be also set by the autoConf feature and not only by UI activity. Renaming the state might be another option but not sure if its used in protobuf as well and if it would have any backward compatibility issues (likely not, but prob. not worth the risk and effort).

chimp1984 commented on core/src/main/java/bisq/core/trade/TradeManager.java:
```diff
@@ -932,65 +935,118 @@ public void processCounterCurrencyExtraData(Trade trade) {
+            return success;
         }
+
+        log.warn("Tx Proof Failure {}, shutting down all open API requests for this trade {}",
```
Can you comment more details when this case is expected?

chimp1984 commented on core/src/main/java/bisq/core/user/Preferences.java:
```diff
@@ -120,6 +120,10 @@
 
+    // list of XMR proof providers : this list will be used if no preference has been set
+    public static final List<String> DEFAULT_XMR_PROOF_PROVIDERS = new ArrayList<> (Arrays.asList(
+            "monero3bec7m26vx6si6qo7q7imlaoz45ot5m2b5z2ppgoooo6jx2rqd.onion"));
```
We should get at least 2 at release time.

chimp1984 commented on core/src/main/java/bisq/core/user/Preferences.java:
```diff
@@ -234,6 +238,9 @@ public void readPersisted() {
             setCryptoCurrencies(CurrencyUtil.getMainCryptoCurrencies());
+            // default values for autoconfirmsettings when persisted payload is empty:
+            prefPayload.setAutoConfirmSettings(new AutoConfirmSettings(
+                    false, 5, 10000000, DEFAULT_XMR_PROOF_PROVIDERS));
```
Instead of 10000000 better use Coin.COIN.value

chimp1984 replied:
This code need to be moved outside the if/else clause. 
```
 if (getAutoConfirmSettings() == null) {
            prefPayload.setAutoConfirmSettings(new AutoConfirmSettings(
                    false, 5, Coin.COIN.value, DEFAULT_XMR_PROOF_PROVIDERS));
        }
```
Otherwise it will be only set for new users but not for those who have already the prefPayload but not the autoConfirmSettings

chimp1984 commented on core/src/main/java/bisq/core/user/PreferencesPayload.java:
```diff
@@ -127,8 +127,8 @@
-    // Added with 1.3.7 false be default
-    private boolean autoConfirmXmr;
+    // Added after 1.3.7
+    private AutoConfirmSettings autoConfirmSettings;
```
Annotate with @Nullable

chimp1984 commented on core/src/main/java/bisq/core/user/PreferencesPayload.java:
```diff
@@ -190,7 +190,7 @@ public Message toProtoMessage() {
                 .setBlockNotifyPort(blockNotifyPort)
                 .setTacAcceptedV120(tacAcceptedV120)
-                .setAutoConfirmXmr(autoConfirmXmr);
+                .setAutoConfirmSettings((protobuf.AutoConfirmSettings) autoConfirmSettings.toProtoMessage());
```
Has to be handled as nullable value: Optional.ofNullable(

chimp1984 commented on core/src/main/java/bisq/core/user/PreferencesPayload.java:
```diff
@@ -279,6 +280,6 @@ public static PreferencesPayload fromProto(protobuf.PreferencesPayload proto, Co
                 proto.getBlockNotifyPort(),
                 proto.getTacAcceptedV120(),
-                proto.getAutoConfirmXmr());
+                AutoConfirmSettings.fromProto(userAutoConfirmSettings));
```
Has to be handled as nullable value

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -569,6 +569,11 @@ portfolio.pending.step2_buyer.startPayment=Start payment
 portfolio.pending.step3_buyer.waitPaymentArrived=Wait until payment arrived
 portfolio.pending.step3_seller.confirmPaymentReceived=Confirm payment received
+portfolio.pending.step3_seller.autoConfirmStatus=Auto-Confirm Status
+portfolio.pending.autoConfirmTxNotFound=Tx Not Found
```
If space in UI permits better use full term `transaction` instead of tx.

chimp1984 replied:
also lower case...
portfolio.pending.autoConfirmTxNotFound=Transaction not found

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -569,6 +569,11 @@ portfolio.pending.step2_buyer.startPayment=Start payment
 portfolio.pending.step2_seller.waitPaymentStarted=Wait until payment has started
 portfolio.pending.step3_buyer.waitPaymentArrived=Wait until payment arrived
 portfolio.pending.step3_seller.confirmPaymentReceived=Confirm payment received
+portfolio.pending.step3_seller.autoConfirmStatus=Auto-Confirm Status
```
portfolio.pending.step3_seller.autoConfirmStatus=Auto-confirm status

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -718,8 +723,8 @@ portfolio.pending.step3_seller.amountToReceive=Amount to receive
 portfolio.pending.step3_seller.yourAccount=Your trading account
-portfolio.pending.step3_seller.xmrTxHash=Tx hash
-portfolio.pending.step3_seller.xmrTxKey=Tx private key
+portfolio.pending.step3_seller.xmrTxHash=Tx id
```
We use `Transaction ID` if space permits, otherwise `Tx ID`

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -1232,6 +1241,9 @@ setting.about.shortcuts.sendFilter=Set Filter (privileged activity)
 
+setting.info.headline=New XMR Autoconfirm feature
+setting.info.msg=When selling BTC for XMR you can use the autoconfirm feature to verify that the correct amount of XMR has been deposited to your wallet and automatically confirm the trade, releasing BTC to the counterparty.\n\n\
+  This can make the trade process more efficient for both peers.  Limits can also be set on the max amount of BTC that can be auto-confirmed.  See the auto-confirm section on this page for more info and to enable the feature.
```
We should request from m52go to review/improve the text.

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -2493,8 +2506,8 @@ showWalletDataWindow.walletData=Wallet data
-setXMRTxKeyWindow.txHash=Transaction hash
-setXMRTxKeyWindow.txKey=Tx private key
+setXMRTxKeyWindow.txHash=Transaction id
+setXMRTxKeyWindow.txKey=Tx key
```
We should stay consistent with the term. I think `Transaction private key` is used in monero.

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -718,8 +723,8 @@ portfolio.pending.step3_seller.amountToReceive=Amount to receive
-portfolio.pending.step3_seller.xmrTxKey=Tx private key
+portfolio.pending.step3_seller.xmrTxHash=Tx id
+portfolio.pending.step3_seller.xmrTxKey=Tx key
 portfolio.pending.step3_seller.xmrTxVerificationError=The XMR transfer validation for trade with ID ''{0}'' failed.\nReason: ''{1}''
```
Seems xmrTxVerificationError is not used anymore

chimp1984 commented on desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/steps/seller/SellerStep3View.java:
```diff
@@ -76,8 +76,8 @@
     private Timer timeoutTimer;
-    private final ChangeListener<XmrProofResultWithTradeId> xmrProofResultWithTradeIdListener;
-
+    TextFieldWithCopyIcon autoConfirmStatusField;
```
should be private

chimp1984 commented on desktop/src/main/java/bisq/desktop/main/settings/preferences/PreferencesView.java:
```diff
@@ -895,9 +915,63 @@ private void activateDaoPreferences() {
+        }
+
+        autoConfirmXmr.setOnAction(e -> {
+            boolean bEnabled = autoConfirmXmr.isSelected();
```
We dont use type prefixes

chimp1984 commented on desktop/src/main/java/bisq/desktop/main/settings/preferences/PreferencesView.java:
```diff
@@ -895,9 +915,63 @@ private void activateDaoPreferences() {
     }
 
+    private void activateAutoConfirmPreferences() {
+        {
```
Why you use an initializer block here?

chimp1984 commented on core/src/main/java/bisq/core/trade/asset/xmr/XmrProofInfo.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    public XmrProofResult checkApiResponse(String jsonTxt) {
+        try {
```
I think it would be useful to log the full jsonTxt.

chimp1984 replied:
Cool!

chimp1984 commented:
@jmacxx Thanks for the work on the feature! Looks pretty good already.

Some comments: 
- I would recommend to use Capitalistion in comments and log messages. Proper sentences makes logs easier to read.  
- Could you merge current master into your branch, should not cause any conflicts.
- Translation strings should be lower case inside sentence...
- Hex validation is done in several places. Would be better to move to Hex util class. Regex is not needed as encode/decode throws if input is wrong. Just need additional lenght check.
- Maybe we should consider to move the processCounterCurrencyExtraData method in TradeManager to a new class to not over burden the TradeManager too much with this specialized feature. It is quite large already...
- We mix a bit between generice feature and XMR specific. Even there are no concrete plans to add the autoconf feature to other altcoins we might add it (e.g. if it get sponsored by the altocin), so better to have all prepared to not get issues when we add more. UI can easily be changed, but data in protobuf, trade, messages are more problematic. 
- In preferences view the auto conf feature is at the very bottom. We  could decrease vertical space for currency selection to make it more prominent or even move it in front of currency selection. 
- The addressRegexValidator does not allow http:// only the IP or onion. 
- Deleting the service address does not refill it with default only at UI screen change or restart. Its a bit tricky to do that. Usually we used a focusOutHandler on the inputfield and apply it when the user leaves focus. 
- The Auto-confirmed badge label is not veritcally centered. Might be a rounding issue with layout...seems to be 1 pixel off.
- Not sure if needed, but might be nice to see the status information at the sellers pending trade view. e.g. how many confirmations from which services. But not sure if UI space allows that. Can be added also later if we see demand for it.
- For dev testing it would be good to deactivate certain checks, like that the key was not used in the past, the date check and the amount check. Otherwise testers need to create a XMR tx for each test run... The DevEnv.devMode property could be used to ignore those checks. Also would be convenient to autofill the buyers tx id and txkey fields with valid data.
- We should consider to allow emty tx ID and tx key input but show a warning if user does not want to send those data. I am not aware of any privacy issue by revealing those data, but it could be that there are some or at least some traders think there are some. Beside that if the users used a wallet which does not provide those data it would lead to a dispute case as he cannot confirm that he sent the xmr. It is mandatory to use a wallet which supports those data, but not sure if we should enforce disputes where otherwise if the seller recieved it there would not be a dispute.
- We should also state in the info popup that it is best for security and privacy that the user runs his own service node and should provide some easy setup instructions (e.g. install script) . beside that we should make clear that there might be reasons why the severice does not provide the result and the user still need to check in the permitted trade time the state of the trade.
- We should add some more detailed info text outside of the app (e.g. blog, wiki, docs). @m52go what do you think? We should make clear the limitations and context of the service. E.g. security is based on assumption that 2 bonded role owners do not collude to defraud traders. There is a privacy loss as the service operators learn about the matching of an onion address and a XMR tx id and amount. All that can be avoided if user runs own service. And of course it is an optional feature if traders dont feel comfortable with it.
- It would be nice to get some statistics how much the feature is used. We could get number of trades using the feature from the node operators (excluding those who run their own service).  
- Once we are in a more final state we should think about the edge cases like when wallet is not synced.... also to define test use cases would help ourself for dev testing and later the testers for release testing.

ripcurlx commented:
The boring bot unfortunately will always tag it as `no-priority` as it touches the altcoin repo. Just ignore this for now as it has `priority`.

m52go commented:
> We should add some more detailed info text outside of the app (e.g. blog, wiki, docs).

Sure, I can write draft points to cover in the next day or two for feedback to make sure material is covered completely and correctly.

ghost commented:
> We mix a bit between generice feature and XMR specific. Even there are no concrete plans to add the autoconf feature to other altcoins we might add it (e.g. if it get sponsored by the altocin), so better to have all prepared to not get issues when we add more. UI can easily be changed, but data in protobuf, trade, messages are more problematic.

Changed proto definition to allow a list of AutoConfirmSettings.
Commit https://github.com/bisq-network/bisq/pull/4421/commits/4dde9e19cf518af5cd2609cf893d0e03f0b29564

> Maybe we should consider to move the processCounterCurrencyExtraData method in TradeManager to a new class to not over burden the TradeManager too much with this specialized feature. It is quite large already...

Moved the new code out to AutoConfirmManager.java
Commit https://github.com/bisq-network/bisq/pull/4421/commits/13978e33c2e0882ebb291f3a713bdc5d612ef500

> Hex validation is done in several places. Would be better to move to Hex util class. Regex is not needed as encode/decode throws if input is wrong. Just need additional lenght check.

Is this referring to the Transaction ID and key input fields?  I should not restrict them to hex input?

> The addressRegexValidator does not allow http:// only the IP or onion.

`address:port` is all that is needed.  Bisq chooses the protocol (http).

> Not sure if needed, but might be nice to see the status information at the sellers pending trade view. e.g. how many confirmations from which services. But not sure if UI space allows that. Can be added also later if we see demand for it.

It shows the number/expected confirmations.
![image](https://user-images.githubusercontent.com/47253594/91088214-39686280-e617-11ea-9c60-d62a00e48ed3.png)

> The Auto-confirmed badge label is not veritcally centered. Might be a rounding issue with layout...seems to be 1 pixel off.

I intentionally oriented the badge slightly above and to the right of the title to make it appear similar to other badges shown on menu items.  I think it looks great that way, but could certainly move it down a bit if that is what you want.

> We should also state in the info popup that it is best for security and privacy that the user runs his own service node and should provide some easy setup instructions (e.g. install script) . beside that we should make clear that there might be reasons why the severice does not provide the result and the user still need to check in the permitted trade time the state of the trade.

Done.  Commit https://github.com/bisq-network/bisq/pull/4421/commits/99146009b6bfdb22e333d11f99c0110436166f71
Will post a screenshot so @m52go can review.

> We should consider to allow emty tx ID and tx key input but show a warning if user does not want to send those data. I am not aware of any privacy issue by revealing those data, but it could be that there are some or at least some traders think there are some. Beside that if the users used a wallet which does not provide those data it would lead to a dispute case as he cannot confirm that he sent the xmr. It is mandatory to use a wallet which supports those data, but not sure if we should enforce disputes where otherwise if the seller recieved it there would not be a dispute.

It allows empty entry; I've also added the warning in that case.  Will post a screenshot so @m52go can review.

> For dev testing it would be good to deactivate certain checks, like that the key was not used in the past, the date check and the amount check. Otherwise testers need to create a XMR tx for each test run... The DevEnv.devMode property could be used to ignore those checks. Also would be convenient to autofill the buyers tx id and txkey fields with valid data.

Done.

> Deleting the service address does not refill it with default only at UI screen change or restart. Its a bit tricky to do that. Usually we used a focusOutHandler on the inputfield and apply it when the user leaves focus.

Done.

> I would recommend to use Capitalistion in comments and log messages. Proper sentences makes logs easier to read.
> Translation strings should be lower case inside sentence...

Duly noted; I have attempted to correct all those instances of mis-capitalization.

ghost commented:
@m52go 

New settings feature text

![image](https://user-images.githubusercontent.com/47253594/91089029-957fb680-e618-11ea-9af5-f9652b38ecd9.png)

Warning when XMR trans keys not provided

![image](https://user-images.githubusercontent.com/47253594/91089288-fd360180-e618-11ea-969b-e1c478015df0.png)

chimp1984 commented:
@jmacxx Thanks for your fast response and implementations! I will try to find time for updated review tomorrow.

chimp1984 commented on assets/src/main/java/bisq/asset/CryptoNoteAddressValidator.java:
```diff
@@ -62,7 +62,7 @@ public AddressValidationResult validate(String address) {
 
     // obtain the raw hex version of a monero address
-    public static String addressToHex(String address) {
+    public static String convert(String address) {
```
if u name method more specific it does not require the comment. e.g. convertToRawHex()

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -642,7 +642,11 @@ portfolio.pending.step2_buyer.fasterPaymentsHolderNameInfo=Some banks might veri
 portfolio.pending.step2_buyer.confirmStart.yes=Yes, I have started the payment
-
+portfolio.pending.step2_buyer.confirmStart.warningTitle=You have not provided proof of payment
+portfolio.pending.step2_buyer.confirmStart.warning=You have not entered the transaction ID and the transaction key.\n\n\
```
Suggestion: portfolio.pending.step2_buyer.confirmStart.warning=You have not entered the transaction ID and the transaction key.\n\n\
  By not providing this data the peer cannot use the auto confirm feature to release the BTC as soon the XMR has been received.\n\
  Beside that, Bisq requires that the sender of the XMR transaction is able to provide this information to the mediator or arbitrator in case of a dispute.

chimp1984 commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -1240,9 +1244,11 @@ setting.about.shortcuts.sendFilter=Set Filter (privileged activity)
+  This can make the trade process more efficient for both peers.  Limits can also be set on the max amount of BTC that can be auto-confirmed.  See the auto-confirm settings on this page for more info and to enable the feature.\n\n\
+  Auto confirmation relies on Bisq checking the XMR transaction on an explorer API.  For maximum security it is recommended that you point Bisq to your own explorer service, but there are also services available run by Bisq (the default setting).\n\n\
+  For more information and details on how to setup, see the blog post here: [.....]
 
```
Suggestions:
setting.info.msg=When selling BTC for XMR you can use the auto confirm feature to verify that the correct amount of XMR has been deposited to your wallet and automatically confirm the XMR receipt and releasing the BTC to the counterparty.\n\n\
  This can make the trade process more efficient for both peers.  Limits can also be set on the max amount of BTC per trade that can be auto-confirmed as well as the number of required confirmations. \n\n\
  Auto confirmation relies on Bisq checking the XMR transaction on an XMR explorer node using the private transaction key provided by the XMR sender. Bisq requests by default from at least 2 nodes operated by Bisq contributors. Both need to provide the same result for triggering the confirmation.
  For maximum security and privacy it is recommended that you run your own XMR explorer node and set that address in the settings.\n\n\
  More information and details on how to setup your own node can be found in the Bisq wiki at: [.....]

@m52go Could you review that?

m52go replied:
This text assumes it will be shown on the settings screen:
```
setting.info.msg=When selling BTC for XMR you can use the auto-confirm feature to verify that the correct amount of 
XMR was sent to your wallet so that Bisq can automatically mark the trade as complete, making trades quicker for everyone.\n\n
Auto-confirm checks the XMR transaction on at least 2 XMR explorer nodes using the private transaction key provided 
by the XMR sender. By default, Bisq uses explorer nodes run by Bisq contributors, but we recommend running your
own XMR explorer node for maximum privacy and security.\n\n
You can also set the maximum amount of BTC per trade to auto-confirm as well as the number of required 
confirmations here in Settings.\n\n
See more details (including how to set up your own explorer node) on the Bisq wiki: [.....]
```
I expect the wiki entry to be fairly short, mainly covering:
- what the auto-confirm feature does
- how to set up an xmr explorer node
- relevant settings: max btc, number of confirmations, and node address

@jmacxx @chimp1984 is there a reference somewhere for how to set up an XMR explorer node?

m52go replied:
Thanks! I've collected what I could in this article:
https://bisq.wiki/Draft:Auto-confirming_trades

It's in the draft namespace so it shouldn't be easily visible just yet.

Feel free to suggest any changes, missing details, etc (something I wondered: how is explorer address found/determined for putting into Bisq?).

m52go replied:
@jmacxx for what it's worth you can probably include the non-draft wiki URL [in the strings](https://github.com/bisq-network/bisq/pull/4421/files#diff-2b16c2ca0e92ce28cd7268b07a143647R1253) for this PR and I'll make it live when it's merged.

https://bisq.wiki/Auto-confirming_trades

chimp1984 reviewed (changes requested):
We need at least 2 providers for DEFAULT_XMR_PROOF_PROVIDERS. Can you push wiz and emzy that they have the updated versions running? One need to run with clearnet.

Clearnet handling of provider for non-tor dev setup is missing.

In Trade:
Do not set xmrProofResult in constructor as it is only used in special circumstances. Keeping it null is more clear. (-> remove xmrProofResult = new XmrProofResult(0, 0, XmrProofResult.State.FEATURE_DISABLED);)
Makr xmrProofResult transient and @Nullable
setXmrProofResult can be package scope.

  	// xmrProofResult is not persisted yet
    @Getter
    @Nullable
    private transient XmrProofResult xmrProofResult;

    void setXmrProofResult(XmrProofResult xmrProofResult) {
        this.xmrProofResult = xmrProofResult;
        xmrProofResultProperty.setValue(xmrProofResult);
    }
    
We should consider to persist XmrProofResult in trade. for trade history it might be important to have some "proof" if autoconfirm was used or not, as well in dispute cases.
if we use propbug for XmrProofResult the enum should not be used as enum in protobuf as protobuf enums are cumbersome (have global scope). use the name instead. see other enum serialisations...(ProtoUtil.enumFromProto)

Still capitalized:
setting.preferences.autoConfirmXMR=XMR Auto Confirm
setting.preferences.autoConfirmEnabled=Enabled
setting.preferences.autoConfirmRequiredConfirmations=Required Confirmations
setting.preferences.autoConfirmMaxTradeSize=Max Trade Size (BTC)
setting.preferences.autoConfirmServiceAddresses=Service Addresses
setting.info.headline=New XMR Auto confirm feature
filterWindow.disableAutoConf=Disable Auto-confirmation (altcoins)


setXMRTxKeyWindow.txHash=Transaction id -> setXMRTxKeyWindow.txHash=Transaction ID
setXMRTxKeyWindow.txKey=Tx key -> setXMRTxKeyWindow.txKey=Transaction key

The second popup at xmr key input is removing the shaded background. You need to start that second popup with a short delay (UserThread.runAfter with 50 ms or so).
Auto confirm settings should be better visible. Either move it on top or reduce number of entries in currency list.

handleProofResult:
This handling need more thought/work:
 if (!walletsSetup.hasSufficientPeersForBroadcast()) {
            return failure;
        }
        if (!walletsSetup.isDownloadComplete()) {
            return failure;
        }

chimp1984 commented:
@m52go Can we add a short description and install help to the wiki?

@Emzy @wiz Can you run the updated verison of the services (with tx date delivered)? one service need to provide clearnet as well (i think emzy has that already) - for dev testing in non-tor setup.
Could you provide a installation script so users can easily install the service themself? I got lots of troubles on Debian with compiling the tool...Seems on Ubuntu its easier...

I think it should be handled as bonded role. We could reuse DATA_RELAY_NODE_OPERATOR as role for that.

Emzy commented:
> @Emzy @wiz Can you run the updated verison of the services (with tx date delivered)? one service need to provide clearnet as well (i think emzy has that already) - for dev testing in non-tor setup.

I just build from master.
If we will use this in production, I need to move the explorer to another machine. It's running on a old pc at home.

ghost commented:
Hi @chimp1984 thank you for taking time to review again.  

> Clearnet handling of provider for non-tor dev setup is missing.

Clearnet in dev mode is by default - no special handling needed.  The SOCKS5 proxy provider is null in dev mode so HttpClient uses clearnet.  To attest to that I've been using emzy's clearnet service for the past few weeks.  The relevant line in HttpClient:
https://github.com/bisq-network/bisq/blob/f45a26ca54b8fcecdc8b071f0f1eca7d042cb4c4/p2p/src/main/java/bisq/network/http/HttpClient.java#L97-L100

If I have misunderstood please correct me.  

I am working on all other points you brought up.  Then I'll merge in latest from master as requested (after squashing my commits).

Cheers.

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -1051,6 +1062,11 @@ setting.preferences.explorer=Bitcoin block explorer
 setting.preferences.explorer.bsq=BSQ block explorer
 setting.preferences.deviation=Max. deviation from market price
 setting.preferences.avoidStandbyMode=Avoid standby mode
+setting.preferences.autoConfirmXMR=XMR Auto Confirm
```
Small but I suggest using a hyphen, e.g. "XMR Auto-Confirm".

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -1228,6 +1244,11 @@ setting.about.shortcuts.sendFilter=Set Filter (privileged activity)
 setting.about.shortcuts.sendPrivateNotification=Send private notification to peer (privileged activity)
 setting.about.shortcuts.sendPrivateNotification.value=Open peer info at avatar or dispute and press: {0}
 
+setting.info.headline=New XMR Auto confirm feature
```
Small but I suggest using a hyphen here too, e.g. "New XMR auto-confirm feature" or "New XMR Auto-Confirm Feature".

Commit:
Merge remote-tracking branch 'origin/master' into add-xmr-tx-key-service-2

chimp1984 commented:
@jmacxx But the onion address is used or at least if you have persisted it before. I will test again later, but when i tested it i got connection to the onion address. We should add the IP address of the clearnet.

ghost commented:
If the user is in non-tor dev mode they should enter clearnet addresses in the settings.  It is not up to Bisq to pick and choose which addresses are to be used.

Commit:
Add XMR proof result status code TRADE_LIMIT_EXCEEDED

ghost commented:
> Once we are in a more final state we should think about the edge cases like when wallet is not synced.

I've moved the precondition checks such that if they are not met then auto-confirmation is not started.


> .... also to define test use cases would help ourself for dev testing and later the testers for release testing.

Here is my current list of test use cases.  I'll also put this into the main topic so it is more visible.

- New user (preferences)
- Auto-confirm success
- Premature manual confirm by user
- XMR Tx Id / key not provided
- XMR Tx not found
- Re-used XMR TxId
- Feature disabled in settings
- Feature disabled by filter
- Transaction limit exceeded
- Invalid amount
- XMR Trade date too old
- XMR Trade date newer than Bisq trade date (ok)
- Bisq shut down during Auto-confirmation procedure
- XMR API service not reachable


TBD:

> It would be nice to get some statistics how much the feature is used. We could get number of trades using the feature from the node operators (excluding those who run their own service).

> We should consider to persist XmrProofResult in trade. for trade history it might be important to have some "proof" if autoconfirm was used or not, as well in dispute cases.
> if we use propbug for XmrProofResult the enum should not be used as enum in protobuf as protobuf enums are cumbersome (have global scope). use the name instead. see other enum serialisations...(ProtoUtil.enumFromProto)

Commit:
Rename XmrProofResult to AutoConfirmResult

chimp1984 commented:
> If the user is in non-tor dev mode they should enter clearnet addresses in the settings. It is not up to Bisq to pick and choose which addresses are to be used.

Its for devs running the regtest setup. It is just more convenient if the clearnet IP is set instead that devs need to find out where to find that...

Commit:
Persist the auto confirm state of Trade object

Auto confirmed trades will be indicated as such on Trade details.

Commit:
Default to clearnet XMR proof providers when in dev mode

ghost commented:
@wiz :kissing:  Any chance you can start to set up and host an XMR proof service? 
https://bisq.wiki/Draft:Auto-confirming_trades#Set_up_your_own_node

chimp1984 commented:
When setting up a XMR account the popup is very large (exceeding the main window). I guess we should shorten the content as well as add information about the auto-confirm feature.

<img width="1098" alt="Screen Shot 2020-08-29 at 17 58 01" src="https://user-images.githubusercontent.com/54558767/91650913-784d4c80-ea4b-11ea-836d-f676a80f64eb.png">

chimp1984 committed:
Merge branch 'master_upstream' into xmr

m52go commented:
I re-did the wiki page and made it more broad so that it is now a general reference for Monero trading (i.e., proving and verifying transactions), but that required me to change the page's title and location. 

@jmacxx this means https://bisq.wiki/Auto-confirming_trades doesn't exist anymore...please change that link to https://bisq.wiki/Trading_Monero#Auto-confirming_trades.

New page will be at https://bisq.wiki/Trading_Monero. See the draft at https://bisq.wiki/Draft:Trading_Monero.

Suggestion for pop-up:

```
account.altcoin.popup.xmr.msg=Trading XMR on Bisq requires that you understand the following requirement:\n\n\
If selling XMR, you must be able to provide the following information to a mediator or arbitrator in case of a dispute:\n\
- the transaction key (Tx Key, Tx Secret Key or Tx Private Key)\n\
- the transaction ID (Tx ID or Tx Hash)\n\
- the destination address (recipient's address)\n\n\
See the wiki for details on where to find this information on popular Monero wallets:\n\
https://bisq.wiki/Trading_Monero#Proving_payments\n\n\
Failure to provide the required transaction data will result in losing disputes.\n\n\
Also note that Bisq now offers automatic confirming for XMR transactions to make trades quicker,\
but you need to enable it in Settings.\n\n\
See the wiki for more about the auto-confirm feature:\n\
https://bisq.wiki/Trading_Monero#Auto-confirming_trades
```

This text skips the sections on verifying transactions and wallet-specific details, but this information is now on the wiki.

For some reason my regtest setup is only letting me add BSQ so I cannot show a graphical rendering of the new text box.

Commit:
Re-word the XMR account creation message per @m52go

ghost commented:
@m52go Done! https://github.com/bisq-network/bisq/pull/4421/commits/d9d0814064c214d79da64ca90a23aeaf7e331360
![image](https://user-images.githubusercontent.com/47253594/91664737-bd5a9880-eab6-11ea-824c-1886f7aef70a.png)

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -1308,35 +1332,18 @@ mediator or arbitrator in case of a dispute.\n\n\
+See the wiki for details on where to find this information on popular Monero wallets:\n\
+https://bisq.wiki/Trading_Monero#Proving_payments\n\n\
+Failure to provide the required transaction data will result in losing disputes.\n\n\
+Also note that Bisq now offers automatic confirming for XMR transactions to make trades quicker,\
```
```suggestion
Also note that Bisq now offers automatic confirming for XMR transactions to make trades quicker, \
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -1308,35 +1332,18 @@ mediator or arbitrator in case of a dispute.\n\n\
-providing this information to the mediator or arbitrator in case of a dispute. In order to do that, \
-you must send XMR using a wallet that provides the information required to prove the payment was made, \
-which includes:\n\n\
+account.altcoin.popup.xmr.msg=Trading XMR on Bisq requires that you understand the following requirement:\n\n\
```
```suggestion
account.altcoin.popup.xmr.msg=Trading XMR on Bisq requires that you understand the following requirement.\n\n\
```

Commit:
Update core/src/main/resources/i18n/displayStrings.properties

Co-authored-by: m52go <mfiver@gmail.com>

Commit:
Update core/src/main/resources/i18n/displayStrings.properties

Co-authored-by: m52go <mfiver@gmail.com>

chimp1984 committed:
- Remove AutoConfirmResult enum from protobuf and add a AutoConfirmResult
instead which stores the stateName. [1]

- Adjust protobuf methods

- Add UNDEFINED to AutoConfirmResult.State to support cases where we
get no data to set the enum.

- Add NO_MATCH_FOUND (used in follow up commits)

- Refactoring: Improve constructors

[1]
Enums in protobuf are not well supported. They are global so an enum
with name (e.g. State) inside Trade conflicts with another enum inside
Message with the same name. So they do not reflect encapsulation in the
class like in java.
We moved over time to the strategy to use strings (from enum.name())
instead of the enum, avoiding also cumbersome fromProto and toProto
code and being more flexible with updates.
The autoConfirmResultState enum inside Trade was a bit confusing to me
as it was a different structure as in the java code. We try to mirror
the structure as far as possible.

chimp1984 committed:
Add null check for name

chimp1984 committed:
- Change tolerance from 1 day to 2 hours.
- Add case if no match is found -> NO_MATCH_FOUND.
- Add test case for NO_MATCH_FOUND
- Add curley brackets to one liners
- Log improvements

chimp1984 committed:
- Make REPEAT_REQUEST_PERIOD and MAX_REQUEST_PERIOD static
- Replace httpClient.toString() with httpClient.getBaseUrl() as toString
would deliver too much for those use cases.
- Add @Nullable
- Log improvements

chimp1984 committed:
- Add @Nullable
- Add curly brackets to one liners
- Log improvements

chimp1984 committed:
Add NO_MATCH_FOUND in comment

chimp1984 committed:
Add abstract AutoConfirmResult class to get better support if we want
to add auto confirm for other currencies in the future. The generic part
is only used where we would have issues with backward compatibility like
in the protobuf objects. Most of the current classes are kept XMR
specific and could be generalized once we add other assets, but that
would be an internal refactoring without breaking any network or
storage data. I think it would be premature to go further here as we
don't know the details of other use cases. I added the methods used from
clients to AutoConfirmResult, not sure if the API is well defined by
that, but as said that could become subject of a future refactoring once
another auto confirm feature gets added. Goal of that refactoring was
to avoid that we need more fields for trade and the the UI would have to
deal with lots of switch cases based on currency.

Sorry that is a larger commit, would have been hard to break up...

chimp1984 committed:
No functionality has been changed by that refactoring, just moved
classes and renamed package and adjusted access modifiers.

chimp1984 committed:
Replace success/failure with booleans

chimp1984 committed:
Apply suggested changes from code analysis

chimp1984 committed:
No functional change, pure refactoring

Move to convertToRawHex CryptoNoteUtils as well as the classes inside
the validator.

chimp1984 commented:
@m52go 
> For some reason my regtest setup is only letting me add BSQ

You need to pay the asset listing fee. luckily on regtest its just virtual money ;-). need  a block conf to get active. so  pay fee for xmr, create a block -> u can use xmr.

chimp1984 committed:
Merge branch 'add-xmr-tx-key-service-2' into add-xmr-tx-key-service-3

chimp1984 committed:
Do not pass over xmrAutoConfirmationManager to ProcessModel but use a
getter from tradeManager. Avoids boilerplate code (I know there is more
in the existing code to optimize here ;-))

chimp1984 committed:
Various small cleanups...

chimp1984 committed:
Refactoring: Rename getTextStatus to getStatusAsDisplayString

chimp1984 committed:
Add @Override, cleanup

chimp1984 committed:
Improve SetXmrTxKeyWindow

chimp1984 committed:
Merge branch 'master_upstream' into add-xmr-tx-key-service-3

chimp1984 committed:
Add devMode check for NO_MATCH case.
Inline method

chimp1984 committed:
Refactoring:
- Move xmrProofInfo.checkApiResponse to XmrProofParser.parse

chimp1984 committed:
Refactoring:
- Rename getKey to getUID
- Rename key to uid

chimp1984 committed:
Refactoring:
Change test package from bisq.core.trade.asset.xmr to bisq.core.trade.autoconf.xmr

chimp1984 committed:
Refactoring:
-Reduce visibility

chimp1984 committed:
Remove @nullable from socks5ProxyProvider param
Remove @Inject from XmrTxProofHttpClient as we pass the
Socks5ProxyProvider anyway

chimp1984 committed:
Refactor:
- Rename processCounterCurrencyExtraData to startRequestTxProofProcess
- Change activeTrades from Stream to List

chimp1984 committed:
Refactor:
- Rename XmrTransferProofRequester to XmrTransferProofRequest

chimp1984 committed:
Refactor:
- Rename AutoConfirmResult to AssetTxProofResult
- Update protobuf entry

chimp1984 committed:
Refactor:
- Rename XmrAutoConfirmResult to XmrTxProofResult

chimp1984 committed:
Refactor:
- Rename XmrProofParser to XmrTxProofParser (and rename test)

chimp1984 committed:
Refactor:
- Rename XmrProofInfo to XmrTxProofModel

chimp1984 committed:
Refactor:
- Rename XmrTransferProofRequest to XmrTxProofRequest

chimp1984 committed:
Refactor:
- Rename XmrTransferProofService to XmrTxProofRequestService

chimp1984 committed:
Refactor:
- Rename XmrAutoConfirmationManager to XmrTxProofService

chimp1984 committed:
Max amount was set to 0.1 BTC, should be 1 BTC

chimp1984 committed:
Use static field for dev test values

chimp1984 committed:
Various refactorings, bug fixes and improvements.

Sorry for the messy commit... its late ;-)

chimp1984 commented:
@jmacxx Made a PR to your branch.... https://github.com/jmacxx/bisq/pull/1

ghost commented:
Thanks @chimp1984, I'm AFK for the next few hours, will look at this when I return.

Commit:
Merge pull request #1 from chimp1984/add-xmr-tx-key-service-3

Autoconfirm XMR trade if tx check was validated by proof service (v3)

ghost commented:
@chimp1984 I attempted to merge your pull request on top of my branch but got stuck in too many merge conflicts.  It seems we have been refactoring the code at the same time.  I think the best option would be to close this PR and you open a new one with your current snapshot of the code.

ghost commented:
I've merged your latest @chimp1984 .  Will test it out.

chimp1984 commented:
Thanks @jmacxx ! I will continue soon as well testing. best to use 2 services for testing and add some artificial delay at the repsonses so u can see how ui behaves... i used same service twice with some small hacks to not ignore it as it has same uid..

I pinged @wiz how his progress is with getting a node. To have 2 nodes is a mandatory requirement for release IMO.

chimp1984 commented:
@jmacxx I tried to do a pR to your branch but it did not work. So I did a new PR to Bisq. https://github.com/bisq-network/bisq/pull/4458

ghost commented:
Closing as replaced by #4458
